src/common/os_thread.h

```diff
@@ -0,0 +1,154 @@
+
+typedef union {
+	long long align;
+	char padding[56];  /* linux: 56 windows: 8 */
+} os_thread_attr_t;
+
+typedef union {
+	long long align;
+	char padding[128];
+} os_cpu_set_t;
+
+typedef volatile int os_spinlock_t; /* XXX: not implemented on windows */
+
+#define OS_ONCE_INIT 0
+void os_cpu_zero(os_cpu_set_t *set);
+void os_cpu_set(size_t cpu, os_cpu_set_t *set);
+
+#ifndef _WIN32
+#define _When_(...)
+#endif
+int os_once(os_once_t *o, void (*func)(void));
+
+int os_tls_key_create(os_tls_key_t *key, void (*destructor)(void *));
+int os_tls_key_delete(os_tls_key_t key);
+int os_tls_set(os_tls_key_t key, const void *value);
+void *os_tls_get(os_tls_key_t key);
+
+int os_mutex_init(os_mutex_t *__restrict mutex);
+int os_mutex_destroy(os_mutex_t *__restrict mutex);
+_When_(return == 0, _Acquires_lock_(mutex->lock))
+int os_mutex_lock(os_mutex_t *__restrict mutex);
+_When_(return == 0, _Acquires_lock_(mutex->lock))
+int os_mutex_trylock(os_mutex_t *__restrict mutex);
+int os_mutex_unlock(os_mutex_t *__restrict mutex);
+
+/* XXX - non POSIX */
+int os_mutex_timedlock(os_mutex_t *__restrict mutex,
+	const struct timespec *abstime);
+
+int os_rwlock_init(os_rwlock_t *__restrict rwlock);
+int os_rwlock_destroy(os_rwlock_t *__restrict rwlock);
+int os_rwlock_rdlock(os_rwlock_t *__restrict rwlock);
+int os_rwlock_wrlock(os_rwlock_t *__restrict rwlock);
+int os_rwlock_tryrdlock(os_rwlock_t *__restrict rwlock);
+_When_(return == 0, _Acquires_exclusive_lock_(rwlock->lock))
+int os_rwlock_trywrlock(os_rwlock_t *__restrict rwlock);
+_When_(rwlock->is_write != 0, _Requires_exclusive_lock_held_(rwlock->lock))
+_When_(rwlock->is_write == 0, _Requires_shared_lock_held_(rwlock->lock))
+int os_rwlock_unlock(os_rwlock_t *__restrict rwlock);
+int os_rwlock_timedrdlock(os_rwlock_t *__restrict rwlock,
+	const struct timespec *abstime);
+int os_rwlock_timedwrlock(os_rwlock_t *__restrict rwlock,
+	const struct timespec *abstime);
+
+int os_spin_init(os_spinlock_t *lock, int pshared);
+int os_spin_destroy(os_spinlock_t *lock);
+int os_spin_lock(os_spinlock_t *lock);
+int os_spin_unlock(os_spinlock_t *lock);
+int os_spin_trylock(os_spinlock_t *lock);
+
+int os_cond_init(os_cond_t *__restrict cond);
+int os_cond_destroy(os_cond_t *__restrict cond);
+int os_cond_broadcast(os_cond_t *__restrict cond);
+int os_cond_signal(os_cond_t *__restrict cond);
+int os_cond_timedwait(os_cond_t *__restrict cond,
+	os_mutex_t *__restrict mutex, const struct timespec *abstime);
+int os_cond_wait(os_cond_t *__restrict cond,
+	os_mutex_t *__restrict mutex);
+
+/* threading */
+
+int os_thread_create(os_thread_t *thread, const os_thread_attr_t *attr,
+	void *(*start_routine)(void *), void *arg);
+
+int os_thread_join(os_thread_t thread, void **result);
+
+/* thread affinity */
+
+int os_thread_setaffinity_np(os_thread_t thread, size_t set_size,
+	const os_cpu_set_t *set);
+
+
+int os_thread_atfork(void (*prepare)(void), void (*parent)(void),
+	void (*child)(void));
+
+
```

Comment:
move `os_semaphore` here, and maybe unify the behavior to also use preallocated memory.